hazelcast/src/main/java/com/hazelcast/datalink/impl/DataLinkConsistencyChecker.java

```diff
@@ -0,0 +1,114 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.datalink.impl;
+
+import com.hazelcast.config.DataLinkConfig;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.datalink.impl.DataLinkServiceImpl.DataLinkSource;
+import com.hazelcast.datalink.impl.DataLinkServiceImpl.DataLinkSourcePair;
+import com.hazelcast.jet.impl.JetServiceBackend;
+import com.hazelcast.map.IMap;
+import com.hazelcast.spi.impl.NodeEngine;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.schema.datalink.DataLink;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DataLinkConsistencyChecker {
+    private final HazelcastInstance hazelcastInstance;
+    private final DataLinkServiceImpl dataLinkService;
+    // sqlCatalog supposed to be initialized lazily
+    private IMap<Object, Object> sqlCatalog;
+
+    private boolean initialized = false;
+
+    public DataLinkConsistencyChecker(HazelcastInstance instance, NodeEngine nodeEngine) {
+        this.hazelcastInstance = instance;
+        this.dataLinkService = (DataLinkServiceImpl) nodeEngine.getDataLinkService();
+    }
+
+    /**
+     * Lazily initialize sqlCatalog.
+     * The only reason of this method existing is
+     * to prevent eager creation of SQL catalog IMap.
+     */
+    public void init() {
+        sqlCatalog = hazelcastInstance.getMap(JetServiceBackend.SQL_CATALOG_MAP_NAME);
+        initialized = true;
+    }
+
+    /**
+     * Removes & alters all outdated data links in __sql.catalog
+     * and adds all missed data links to __sql.catalog
+     */
+    public void check() {
+        if (!initialized) {
+            return;
+        }
+
+        List<Map.Entry<String, DataLinkSourcePair>> sqlEntries = dataLinkService.getDataLinks()
+                .entrySet()
+                .stream()
+                .filter(en -> en.getValue().source == DataLinkSource.SQL)
+                .collect(Collectors.toList());
+
+        for (Map.Entry<String, DataLinkSourcePair> entry : sqlEntries) {
+            Object catalogValue = sqlCatalog.get(QueryUtils.wrapDataLinkKey(entry.getKey()));
+
+            // Data link is absent in sql catalog -> add it.
+            if (catalogValue == null) {
+                addEntryToCatalog(entry);
+            } else {
+                // Try to alter outdated data links
+                DataLink catalogDataLink = (DataLink) catalogValue;
+                if (!dataLinksAreEqual(entry.getValue(), catalogDataLink)) {
+                    sqlCatalog.remove(QueryUtils.wrapDataLinkKey(entry.getKey()));
+                    addEntryToCatalog(entry);
+                }
+            }
+        }
+
+        // Data link is not present in service, but still present in sql.catalog
+        sqlCatalog.removeAll(e -> e.getValue() instanceof DataLink
```

Comment:
There can be a race between removing links from `sqlCatalog` (`IMap`) and adding them on another node: local `dataLinkService` may not yet see newly added link. I could not find if the datalinks are propagated eagerly to local caches on other nodes.

Reply:
I know about it and since this call is constantly repeated, I see no big issue with the current approach. On the next call (it happens once per second) system will try to synchronize state again.

Reply:
where do we replicate removals of the data links? `DROP DATA LINK` only updates `DataLinkService` on single node. Other nodes will see it missing in the SQL catalog but present in their local copy so they will add such link again.